CMakeLists.txt

```diff
@@ -174,9 +174,47 @@ foreach(src_file ${uhdm-GENERATED_SRC})
   set_source_files_properties(${src_file} PROPERTIES GENERATED TRUE)
 endforeach(src_file ${uhdm-GENERATED_SRC})
 
+# Parse out version and configure header
+if(NOT DEFINED ENV{UHDM_VERSION_MAJOR})
+  # parse from working directory
+  # ensure `git` is installed
+  find_package (Git REQUIRED)
+
+  # get latest commit
+  execute_process(COMMAND git describe --tags --abbrev=0
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    OUTPUT_VARIABLE LATEST_VERSION)
+
+  # parse out first number after "v"
+  string(REGEX REPLACE "^v([0-9]+)\\..*" "\\1" UHDM_VERSION_MAJOR "${LATEST_VERSION}")
+  # parse out second number, e.g. after "v[0-9]."
+  string(REGEX REPLACE "^v[0-9]+\\.([0-9]+).*" "\\1" UHDM_VERSION_MINOR "${LATEST_VERSION}")
+
+  # get latest commit
+  execute_process(COMMAND git rev-parse HEAD
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    OUTPUT_VARIABLE UHDM_VERSION_COMMIT_SHA)
+  # strip newline
+  string(REGEX REPLACE "\n$" "" UHDM_VERSION_COMMIT_SHA "${UHDM_VERSION_COMMIT_SHA}")
+else()
+  # if set in env, assume some other deployment mechanism where
+  # not working from the git repo / where git not installed
+  set(UHDM_VERSION_MAJOR ENV{UHDM_VERSION_MAJOR})
```

Comment:
What happens if a user just downloads the source archive and compiles ? It looks like in that case they might need to set the environment variables correctly.

Can we set the reasonable defaults in the CMakeFile directly ? Maybe providing a version directly in the [`project()` cmake](https://cmake.org/cmake/help/latest/command/project.html) function (around line 26)

```cmake
project(UHDM VERSION 1.48)
```

Now if `git` is available, we still do the extracting from the git repo with the exact git tag and sha hash.
(we could generate a warning relevant for developers if the tag returns a different major/minor than we expect in the project version).

If git does not exist, we configure the `UHDM_VERSION_MAJOR`, `UHDM_VERSION_MINOR` which is automatically set by the `project()` call. and set the `UHDM_VERSION_COMMIT_SHA` to `"unknown"` or something similar.

Reply:
Yes, cmake version should be source of truth.
The current tagging process (which still is manually) also uses the source to deduce the version number (but only in Surelog currently).

Goal is to have the tag happening automatically at CI time, never manually.

Reply:
sounds good